### PR TITLE
Add _translate_constraint_by_config function

### PR DESCRIPTION
**Pull request by skycastlelily**
Pull Request Checklist

* [ ] implement the feature
* [ ] write the documentation
* [ ] extend the test coverage
* [ ] update the specification
* [ ] adjust plugin docstring
* [ ] modify the json schema
* [ ] mention the version
* [ ] include a release note

### CODE REVIEW

**Comment by skycastlelily** [comment]
This MR looks at config file and tries to find a "translation" that would match the constraint, operator, value, and if found, it'd take it from configuration^^

File: tmt/steps/provision/mrack.py [hunk 1]
```diff
@@ -40,6 +42,7 @@
 
 # Type annotation for "data" package describing a guest instance. Passed
 # between load() and save() calls
+MRACK_CONFIG_FILE = os.environ.get('MRACK_CONFIG_FILE') or 'mrack-hw-filters.yaml'
```
Comment (happz) [inline tmt/steps/provision/mrack.py:45]: 1. comment is now separated from the structures it's commenting by this addition, I suppose this particular line should move before the command.
2. `os.environ.get()` supports default value, i.e. `os.environ.get('MRACK_CONFIG_FILE', 'mrack-hw-filters.yaml')` would work as designed.
3. the environment variable says `CONFIG_FILE` which is not exactly true - it's not a config file for mrack, just for one of the steps. And the default filename proves that, it's not `mrack-config.yaml`, but much better `mrack-hw-filters.yaml`. So how about `MRACK_HW_FILTERS_CONFIG_FILE` or something similar?
4. all environment variables accepted by tmt should have `TMT_` prefix, see https://tmt.readthedocs.io/en/stable/overview.html#variables. We cannot let tmt to conflict with other apps.

File: tmt/steps/provision/mrack.py [hunk 2]
```diff
@@ -197,6 +202,34 @@ class MrackHWNotGroup(MrackHWGroup):
     name: str = 'not'
 
 
+def _get_custom_config() -> dict[str, Any]:
```
Comment (happz) [inline tmt/steps/provision/mrack.py:205]: This function does not seem to be called from anywhere.

File: tmt/steps/provision/mrack.py (hunk 1 above)
Comment (happz) [reply tmt/steps/provision/mrack.py:45]: It would be nice to mention the config file and its structure in the plugin help, i.e. in its docstring, with an example or two. It does not have to do much now, but it would help other developers to understand what and why - I know it, because it comes from Artemis, but others might be lost during reviews. And then there are users :)

**Comment by skycastlelily** [comment]
yeah,it will be used if any _transform_*_*_  function calls
_translate_constraint_by_config,should I move this function to a new merge
request?^^

**Comment by skycastlelily** [comment]
Updated^^

File: tmt/steps/provision/mrack.py [hunk 3]
```diff
@@ -197,6 +206,34 @@ class MrackHWNotGroup(MrackHWGroup):
...
+    if TMT_MRACK_HW_FILTERS_FILE.is_absolute():
+        config_content = TMT_MRACK_HW_FILTERS_FILE.read_text()
+    else:
+        config_content = (tmt.utils.Config().path / TMT_MRACK_HW_FILTERS_FILE).read_text()
+    return tmt.utils.yaml_to_dict(config_content)
```
Comment (happz) [inline tmt/steps/provision/mrack.py:214]: Right, I would not let user override just this particular file when there's a nice envvar for defining where tmt configuration lives. So this should always use `Config.path`, plus the file name mentioned above.

File: tmt/steps/provision/mrack.py [hunk 4]
```diff
@@ -197,6 +206,34 @@ class MrackHWNotGroup(MrackHWGroup):
...
+
+
+def _translate_constraint_by_config(
+        constraint: tmt.hardware.Constraint[Any],
+        translations: list[Any]
+        ) -> dict[str, Any]:
+    for translation in translations:
```
Comment (happz) [inline tmt/steps/provision/mrack.py:221]: It would be extremely useful to include an example config file. It should be small, well-documented, showing an example or two - commented out if needed. Right now, all we have is guessing based on the unit test below.

Another point I'm not sure about: what should the file contain? XML, or the simplified node-like mappings we have above? Or something else? E.g. we have something like this in Artemis:

```yaml
  compatible:
    distro:
      translations:
        - operator: contains 
          value: rhel-6
          element: '<compatible_with_distro arch="<< ENVIRONMENT.hw.arch >>" osmajor="RedHatEnterpriseLinux6"/>'
```

I.e. maps `compatible.distro` constraint to an actual XML. It's needlessly complicated, we could have made it simpler, but there needs to be a selector and the pattern, and I'm thinking about generating the actual XML, with Jinja because the use of templates proved to be very valuable - without templates, it'd be very hard to inject guest architecture into the element above. Simplified, it might be something like this (with `GUEST` representing the tmt guest for which we are generating the filter):

```yaml
compatible.distro: |
  {% if GUEST.image | regex_match('(?i)rhel-6') %}
    <compatible_with_distro arch="{{ GUEST.arch }}" osmajor="RedHatEnterpriseLinux6"/>'
  {% endif %}
```

We could get rid of `operator` and `translation` and leave it to Jinja. `operator` as a concept did not prove to be very flexible in Artemis implementation - this sucks:

```yaml
    is_virtualized:
      translations:
        - operator: '=='
          value: false
          element: '<system><hypervisor op="==" value="" /></system>'

        - operator: '=='
          value: true
          element: '<system><hypervisor op="!=" value="" /></system>'
```

What I don't know is how to take the output of such a template, and connect it with the tree or dataclasses we build. Dictionaries would be easier, and you took that route, but those don't make the use of templates easier:

```yaml
compatible.distro:
  compatible_with_distro:
    arch: "{{ GUEST.arch }}"
    osmajor: {% if GUEST.image | regex_match('(?i)rhel-6') %}RedHatEnterpriseLinux6{% endif %}

# Or, the value is nothing but a template string, we then `yaml_to_dict()` it once more:
compatible.distro: |
  compatible_with_distro:
    arch: "{{ GUEST.arch }}"
    osmajor: {% if GUEST.image | regex_match('(?i)rhel-6') %}RedHatEnterpriseLinux6{% endif %}
```

In other words, how can we make the filter parametrized so we wouldn't have to enumerate every single distro and arch combination for `compatible.distro`? :)

File: tmt/steps/provision/mrack.py (hunk 3 above)
Comment (skycastlelily) [reply tmt/steps/provision/mrack.py:214]: Got it, will update accordingly:)

**Comment by skycastlelily** [comment]
Sorry for the a bit late reply, I took half day pto today:

> It would be extremely useful to include an example config file

 sure

>  I'm not sure about: what should the file contain?

something similar with  the one you said sucks 😅

  disk:
    model-name:
      translations:
        - operator: ==
          value: PERC H310
          element: {'disk': {'model': {'_op': '==', '_value': 'PERC H310'}}}


> We could get rid of operator and translation and leave it to Jinja
> so we wouldn't have to enumerate every single distro and arch combination
for compatible.distro? :)


I'm  really not sure what I'm going to put there is correct, or I
understand you correctly(Believe me,I try very hard).

For simple ones like I mentioned above it's easy,we can  get rid of
operator (I'm not sure how can we get rid of translation), replace it with
something like:

  disk:
    model-name:
     translations:
     value: PERC H310
    element: {'disk': {'model': {'_op': {{operator}}, '_value': 'DUMMY' }}}

but I think the original reason why we add translations is because they are
complicated ones, for instance:
https://gitlab.cee.redhat.com/baseos-qe/ansible-baseos-ci/-/blob/main/group_vars/artemis.yaml#L1473,we
need value and operator^^

File: tmt/steps/provision/mrack.py (hunk 4 above)
Comment (skycastlelily) [reply tmt/steps/provision/mrack.py]: > We could get rid of operator and translation

We could also get rid of element^^:  https://github.com/teemtee/tmt/commit/1a89ca31bcd54d2ae85129971a42bfd84efb554d

For simple ones,we could use a common  function:_translate_constraint_by_config .
We need to remove the following code from [mack.py](https://github.com/teemtee/tmt/blob/main/tmt/steps/provision/mrack.py#L1449)
```
        if data.hardware:
            data.hardware.report_support(
                names=list(_CONSTRAINT_TRANSFORMERS.keys()), logger=self._logger
            )
```
Plus,we use _transform_unsupported to warn users for "hardware will have no effect", kind of duplicate.
And, it feels like a dedicate mr, so I didn't add that code to this mr:)

For complex ones we could use corresponding functions like we do in [boot-method mr](https://github.com/teemtee/tmt/pull/2671).

File: tmt/steps/provision/mrack.py [hunk 5]
```diff
@@ -63,6 +65,28 @@ def mrack_constructs_ks_pre() -> bool:
...
+    try:
+        hardware_config = tmt.config.Config().hardware
+    except tmt.utils.SpecificationError:
+        raise
+    except tmt.utils.MetadataError:
+        return None
+    if not hardware_config:
+        return None
+
+    return hardware_config.hardware_config
```
Comment (happz) [inline tmt/steps/provision/mrack.py:87]: * I'd recommend returning an empty list, it's as false-ish as `None`
* is it necessary to capture there exceptions, and then either reraise, or silence it? `MetadataError` should be heard as loudly as `SpecificationError`.

I think this could be simplified (plus "custom config" can mean anything, we're speaking about "translations" of constraints directly to Beaker XML):

```python
def _get_constraint_translations() -> list[Translate]:
    config = tmt.config.Config()

    return config.hardware.translations if config.hardware else []
```

File: tmt/config/models/hardware.py [hunk 6]
```diff
@@ -0,0 +1,13 @@
...
+
+
+class Translate(MetadataContainer):
+    name: str
+    child_name: Optional[str]
+    element: str
+
+
+class HardwareConfig(MetadataContainer):
+    hardware_config: list[Translate]
```
Comment (happz) [inline tmt/config/models/hardware.py:13]: I'd probably call it "translations" or something. `Config.hardware.hardware_config` is really puzzling.

File: tmt/steps/provision/mrack.py [hunk 7]
```diff
@@ -236,6 +271,29 @@ def _transform_unsupported(
...
+        ):
+            config_str = config.element
+            break
+        continue
+
+    beaker_operator, actual_value, negate = operator_to_beaker_op(
+        constraint.operator, constraint.value
+    )
+    if config_str:
+        return yaml_to_dict(config_str % (beaker_operator, actual_value))
```
Comment (happz) [inline tmt/steps/provision/mrack.py:293]: Nope, no `%`, no `format()`: tmt should stick to one templating option, this should be `render_template(config_str, ...)`.

File: tmt/steps/provision/mrack.py [hunk 8]
```diff
@@ -236,6 +271,29 @@ def _transform_unsupported(
...
+) -> Union[dict[str, Any], MrackBaseHWElement]:
+    name, _, child_name = constraint.expand_name()
+    config_str = None
+    for config in hardware_config:
+        if (name == config.name and child_name == config.child_name) or (
+            name == config.name and not child_name
+        ):
+            config_str = config.element
+            break
+        continue
```
Comment (happz) [inline tmt/steps/provision/mrack.py:287]: * `hardware_config` => `constraint_translations` or something.
* plus:

```python
suitable_translations = [
    translation
    for translation in constraint_translations
    if translation.name == name and (translation.child_name is None or translation.child_name == child_name)
]

if not suitable_translations:
    return _transform_unsupported(constraint, logger)
```

File: tmt/config/models/hardware.py [hunk 9]
```diff
@@ -0,0 +1,13 @@
+from typing import Optional
+
+from tmt.container import MetadataContainer
+
+
+class Translate(MetadataContainer):
+    name: str
+    child_name: Optional[str]
+    element: str
```
Comment (happz) [inline tmt/config/models/hardware.py:9]: Check out https://github.com/teemtee/tmt/pull/2967#discussion_r1771508347, the second example: we could replace `name` and `child_name` with just `name`, and then use `constraint.printable_name` to compare against. That would compare names like `compatible.distro` or `cpu.processors`, which would be readable.

Also, I'd change `element` to something like `element_template` or just `template`.

File: tmt/steps/provision/mrack.py (hunk 5 above)
Comment (skycastlelily) [reply tmt/steps/provision/mrack.py]: >is it necessary to capture there exceptions, and then either reraise, or silence it? MetadataError should be heard as loudly as SpecificationError

We don't want to raise an error if there is [no config tree](https://github.com/teemtee/tmt/blob/main/tmt/config/__init__.py#L89
),right ? so ,they should be handled differently.WDYT? :)

**Comment by skycastlelily** [comment]
Thanks for your super quick response,and pretty valuable suggestions, updated accordingly^^

File: tmt/config/models/hardware.py [hunk 10]
```diff
@@ -0,0 +1,10 @@
+from tmt.container import MetadataContainer
+
+
+class Translate(MetadataContainer):
+    name: str
```
Comment (happz) [inline tmt/config/models/hardware.py:5]: Hmhm, after some thinking, `name` of what? How about calling this `requirement`, `constraint`, `requirement_name` or something like that? Any ideas? Because it's not supposed to be the name of this translation, but the HW requirement name, label or whatever.

File: tmt/steps/provision/mrack.py [hunk 11]
```diff
@@ -63,6 +65,23 @@ def mrack_constructs_ks_pre() -> bool:
...
 
 
+def _get_constraint_translations() -> list[Translate]:
+    """
+    Read yaml config file and translate its content to a list of Translate.
+    example:
+    translations:
+      - name : cpu.processors
+        template: '{"cpu": {"processors": {"_op": "operator", "_value": "value"}}}'
+    """
```
Comment (happz) [inline tmt/steps/provision/mrack.py:75]: Let's take the advantage of Sphinx, and get some nice docs :)

* the example belongs to `tmt.config.models.hardware`, to one of the classes defined there, as they define the fields and should provide examples of their own use.
* ```python
  """
  Load the list of hardware requirement translations from configuration.

  :returns: translations loaded from configuration, or an empty list if
      the configuration is missing.
  """
  ```

File: tmt/steps/provision/mrack.py [hunk 12]
```diff
@@ -186,15 +205,26 @@ class MrackHWGroup(MrackBaseHWElement):
...
 
-        return {self.name: [child.to_mrack() for child in self.children]}
+        child_list = []
+        for child in self.children:
+            # There is no need to call to_mrack if child is already a dict
+            if isinstance(child, dict):
+                child_list.append(child)
+            else:
+                child_list.append(child.to_mrack())
+        return {self.name: child_list}
```
Comment (happz) [inline tmt/steps/provision/mrack.py:227]: We can still return the dictionary filled through list comprehension, without any extra variables, and I'd say it's still nicely readable:

```python
return {
    self.name: [
        child.to_mrack() if isinstance(child, MrackBaseHWElement) else child
    ]
}
```

File: tmt/steps/provision/mrack.py [hunk 13]
```diff
@@ -676,7 +728,7 @@ def _transform_system_vendor_name(
 
 def constraint_to_beaker_filter(
     constraint: tmt.hardware.BaseConstraint, logger: tmt.log.Logger
-) -> MrackBaseHWElement:
+) -> Union[MrackBaseHWElement, dict[str, Any]]:
```
Comment (happz) [inline tmt/steps/provision/mrack.py:731]: Right, we should have a type alias for these two, a type representing things that a constraint can be translated to, e.g. `BeakerizedConstraint = Union[MrackBaseHWElement, dict[str, Any]]`.

Comment (skycastlelily) [reply tmt/steps/provision/mrack.py]: updated^^:https://github.com/teemtee/tmt/pull/2967/commits/3ac72a44d0d464dbf9392b0fb7d5f34bb91b0e34

File: tmt/steps/provision/mrack.py (hunk 12 above)
Comment (skycastlelily) [reply tmt/steps/provision/mrack.py]: updated^^ :https://github.com/teemtee/tmt/pull/2967/commits/3ac72a44d0d464dbf9392b0fb7d5f34bb91b0e34

File: tmt/steps/provision/mrack.py (hunk 11 above)
Comment (skycastlelily) [reply tmt/steps/provision/mrack.py:74]: updated^^ :https://github.com/teemtee/tmt/pull/2967/commits/3ac72a44d0d464dbf9392b0fb7d5f34bb91b0e34

File: tmt/config/models/hardware.py (hunk 10 above)
Comment (skycastlelily) [reply tmt/config/models/hardware.py]: updated^^ :https://github.com/teemtee/tmt/pull/2967/commits/3ac72a44d0d464dbf9392b0fb7d5f34bb91b0e34

File: tmt/config/models/hardware.py [hunk 14]
```diff
@@ -0,0 +1,10 @@
...
+    template: str
+
+
+class HardwareConfig(MetadataContainer):
+    translations: list[Translate]
```
[the remark this answers is not in the document]
Comment (skycastlelily) [reply tmt/config/models/hardware.py]: updated^^ :https://github.com/teemtee/tmt/pull/2967/commits/3ac72a44d0d464dbf9392b0fb7d5f34bb91b0e34

**Comment by happz** [comment]
Adding `blocked`, to get https://github.com/teemtee/tmt/pull/3519 merged first as it simplifies some pieces of configuration handling.

File: tmt/steps/provision/mrack.py [hunk 15]
```diff
@@ -63,6 +65,22 @@ def mrack_constructs_ks_pre() -> bool:
...
+    :returns: translations loaded from configuration, or an empty list if
+        the configuration is missing.
+    """
+    try:
+        config = tmt.config.Config()
+    except tmt.utils.MetadataError:
+        raise
+    if config.hardware:
+        return config.hardware.beaker.translations if config.hardware.beaker else []
+    return []
```
Comment (happz) [inline tmt/steps/provision/mrack.py:81]: * `MetadataError` should no longer be raised by `tmt.config.Config()`.
* `return` can be combined, with a slightly longer test:
  ```python
  return config.hardware.beaker.translations if config.hardware and config.hardware.beaker else []
  ```

File: tmt/steps/provision/mrack.py [hunk 16]
```diff
@@ -236,6 +266,33 @@ def _transform_unsupported(
     return MrackHWOrGroup()
 
 
+def _translate_constraint_by_config(
+    constraint: tmt.hardware.Constraint[Any],
+    constraint_translations: list[MrackTranslation],
+    logger: tmt.log.Logger,
+) -> BeakerizedConstraint:
```
Comment (happz) [inline tmt/steps/provision/mrack.py:273]: A docstring would be nice.

**Comment by happz** [comment]
BTW the PR deserves a better subject and message.

File: tmt/steps/provision/mrack.py (hunk 15 above)
Comment (skycastlelily) [reply tmt/steps/provision/mrack.py]: wow, impressive again^^: https://github.com/teemtee/tmt/pull/2967/commits/67fd1322d3619d7b05916944025320129b32bfe2

File: tmt/steps/provision/mrack.py (hunk 16 above)
Comment (skycastlelily) [reply tmt/steps/provision/mrack.py:270]: updated:https://github.com/teemtee/tmt/pull/2967/commits/67fd1322d3619d7b05916944025320129b32bfe2

**Review by happz — APPROVED** [review]

**Review by psss — COMMENTED** [review]
Thanks for working on this. Overall looks good. I'm proposing a small enhancement in c20edbd6. Does it look ok? Plus have a few questions:

* The pull request check list states that test coverage has been extended, but there seem to be no changes to tests. What about adding a simple test exercising just `tmt provision --dry` to ensure everything works as expected?
* Just to clarify the motivation: So users will be able to configure (and maintain) their own hardware translation rules? Do we expect this will be a common use case? Do we expect something similar for other provision plugins as well?
* I tried a simple modification of your config example but it does not seem to be working as expected. Am I doing anything wrong?

Config file `.config/tmt/hardware.fmf`:

```yaml
beaker:
    translations:
      - requirement: gpu.processors
        template: '{"gpu": {"processors": {"_op": "{{ OPERATOR }}", "_value": "{{ VALUE }}"}}}'
```

Simple `plan.fmf`:

```yaml
provision:
    how: beaker
    hardware:
        gpu:
            processors: 8
execute:
    how: tmt
```

Output of `tmt run provision --dry`:

```xml
<job retention_tag="audit" product="[internal]">
  <whiteboard>tmt-105-bdBTtPlH</whiteboard>
  <recipeSet priority="Normal">
    <recipe whiteboard="" ks_meta="">
      <distroRequires>
        <and>
          <distro_name op="like" value="Fedora-41%"/>
          <distro_variant op="=" value="Server"/>
          <distro_arch op="=" value="x86_64"/>
        </and>
      </distroRequires>
      <hostRequires>
        <and>
          <and/>
        </and>
      </hostRequires>
      <repos/>
      <partitions/>
      <reservesys duration="86400"/>
      <task name="/distribution/dummy" role="STANDALONE">
        <params/>
      </task>
    </recipe>
  </recipeSet>
</job>
```

Seems that `hostRequires` are not set as expected.

**Comment by skycastlelily** [comment]
> Thanks for working on this.

 Pleasure, thanks for your review.

> I'm proposing a small enhancement in https://github.com/teemtee/tmt/commit/c20edbd6b4f7f003cd9b492d01db607bba422996. Does it look ok?

Yeah, thanks:)

> The pull request check list states that test coverage has been extended, but there seem to be no changes to tests. What >about adding a simple test exercising just tmt provision --dry to ensure everything works as expected?

I checked that list by mistake,sorry for the confusing, do you see where we could add the test? or, maybe you 
could extend the test for us if you are not too busy? :saluting_face: :grin:   You should know we need a hardware  
translated by hardware.py  but not translate by mrack.py, for instance, boot.method. Because if the constraint has a 
[_CONSTRAINT_TRANSFORMERS](https://github.com/teemtee/tmt/blob/main/tmt/steps/provision/mrack.py#L645),
the _translate_constraint_by_config method will not be called. 

Ah, wait, after think twice, maybe we should let custom config surpass the default one in mrack.py,
ie ,we check whether if there is specific  translates in config file first, if not,  use _CONSTRAINT_TRANSFORMERS.
WDYT?:) @happz 

> Just to clarify the motivation: So users will be able to configure (and maintain) their own hardware translation rules? 

 yep

> Do we expect this will be a common use case? Do we expect something similar for other provision plugins as well?

  I guess both of answer is "yes" and @happz may be able to offer more info:)

 > I tried a simple modification of your config example but it does not seem to be working as expected. Am I doing >anything wrong?

That is because hardware.py does not support translate gpu.processors  to hardware constraint ,you need add processors to [number_constraints](https://github.com/teemtee/tmt/blob/main/tmt/hardware.py#L1111). @happz mentor, I remembered you add the gpu translation and  _parse_device_core ,do you think we should add  support processors and maybe frequency for gpu? Smells like a mr for me;
You also need to change VALUE to BEAKER_VALUE in hardware.fmf to have it fully work.

**Comment by psss** [comment]
> Ah, wait, after think twice, maybe we should let custom config surpass the default one in mrack.py,

Yeah, this sounds reasonable to me: To allow users override default translation with their custom settings.

> That is because hardware.py does not support gpu.processors ,you need add processors to [number_constraints](https://github.com/teemtee/tmt/blob/main/tmt/hardware.py#L1111).

So there are additional actions needed from the user to enable a specific translation?

> I checked that list by mistake,sorry for the confusing, do you see where we could add the test? or, maybe you
> could extend the test for us if you are not too busy?

Sorry, no extra time as for now. But I'd say a very simple translation example and check with `provision --dry` would be completely sufficient.

**Comment by happz** [comment]
> > Ah, wait, after think twice, maybe we should let custom config surpass the default one in mrack.py,
> 
> Yeah, this sounds reasonable to me: To allow users override default translation with their custom settings.

It wasn't meant to step in before the hardcoded ones, but rather to close the gap where the hardcoded one does not exist yet or never should, but I think it makes perfect sense to inspect the translations first, then fall back to hardcoded translations, then give up.

> > That is because hardware.py does not support gpu.processors ,you need add processors to [number_constraints](https://github.com/teemtee/tmt/blob/main/tmt/hardware.py#L1111).
> 
> So there are additional actions needed from the user to enable a specific translation?

Yes and no. You tried to translate a HW requirement that is not yet defined by tmt specification. As such, it is not parsed by tmt, and it is not delivered to `beaker` plugin to inspect and convert into Beaker XML. So in this sense, yes, there are additional actions needed, define the constraint and add its support to `tmt.hardware`.

For already-defined constraints, there should be no other action needed. The goal is to give `beaker` a chance to use more complicated, deployment-specific XML snippets instead of hardcoded ones - our current downstream filter for `virtualization.is-virtualized` is many, many lines long and depends on our Beaker's setup. So, `gpu.processors` no, `gpu.driver` yes, that should manifest the functionality.

**Comment by happz** [comment]
> > Just to clarify the motivation: So users will be able to configure (and maintain) their own hardware translation rules?
> 
> yep
> 
> > Do we expect this will be a common use case? Do we expect something similar for other provision plugins as well?
> 
> I guess both of answer is "yes" and @happz may be able to offer more info:)

Yes, and yes: users - and this means also TF - should be able to add their own custom translations to be used with Beaker. Some of them are very complicated, depend on the current setup of our Beaker deployment and it just does not feel right to hardcode these. So the idea, taken from Artemis, is to allow configuration to supply anything that goes beyond trivial translation of requirements like `cpu.cores`.

I do expect something similar for other plugins: Artemis works a lot with instance-type-based clouds, and while it's not the case of `artemis` plugin, `minute` is a prime example of this feature: what OpenStack flavor should be picked for `--how minute --memory '>= 4GB'`? Artemis keeps a list of instance types and their properties, updated dynamically, for `minute` and other plugins we may add in the future I envision such a list to exist in configuration. Just like we have a config file for `mrack`, with distros and distro variants, for example. Our `qe-tools-workstation` can easily ship something like `~/.config/tmt/hardware/minute/main.fmf` with a list of available flavors.

> > I tried a simple modification of your config example but it does not seem to be working as expected. Am I doing >anything wrong?
> 
> That is because hardware.py does not support gpu.processors ,you need add processors to [number_constraints](https://github.com/teemtee/tmt/blob/main/tmt/hardware.py#L1111). @happz mentor, I remembered you add the gpu translation and _parse_device_core ,do you think we should add support processors and maybe frequency for gpu? Smells like a mr for me; You also need to change VALUE to BEAKER_VALUE in hardware.fmf to have it fully work.

Nope, no need to add `gpu.processors`, I believe this is unfortunately a wrong example, no plugin supports `gpu.processors`. `gpu.driver` would be better, as it's already defined by the specification.

**Comment by psss** [comment]
Thanks for the detailed clarification. I've shortly tested with `gpu.driver` and it seems to be working:

This config:

```yaml
beaker:
    translations:
      - requirement: gpu.driver
        template: '{"gpu": {"driver": {{ VALUE }} }}'
```

Together with the plan:

```yaml
provision:
    how: beaker
    hardware:
        gpu:
            driver: something
execute:
    how: tmt
```

Gives this:

```xml
<hostRequires>
  <and>
    <gpu>
      <driver/>
    </gpu>
  </and>
</hostRequires>
```

Any quick hint what the template syntax should look like to get the value into the resulting `xml`?

**Comment by happz** [comment]
> Thanks for the detailed clarification. I've shortly tested with `gpu.driver` and it seems to be working:
> 
> This config:
> 
> ```yaml
> beaker:
>     translations:
>       - requirement: gpu.driver
>         template: '{"gpu": {"driver": {{ VALUE }} }}'
> ```
> 
> Together with the plan:
> 
> ```yaml
> provision:
>     how: beaker
>     hardware:
>         gpu:
>             driver: something
> execute:
>     how: tmt
> ```
> 
> Gives this:
> 
> ```
> <hostRequires>
>   <and>
>     <gpu>
>       <driver/>
>     </gpu>
>   </and>
> </hostRequires>
> ```
> 
> Any quick hint what the template syntax should look like to get the value into the resulting `xml`?

`BEAKER_VALUE` instead of `VALUE`.

I hope to get the documentation in the "generate docs for config from models" batch of PRs, the syntax and examples should be coming from that place.

**Comment by psss** [comment]
> I hope to get the documentation in the "generate docs for config from models" batch of PRs, the syntax and examples should be coming from that place.

Great, thanks!

**Review by psss — APPROVED** [review]

**Comment by skycastlelily** [comment]
Hi Petr, did you  merge this by mistake? :)

I guess by "but I think it makes perfect sense to inspect the translations first, then fall back to hardcoded translations, then give up." Milos has the same opinion with you, ie,  "let custom config surpass the default one in mrack.py"
I already modify the code accordingly and halfway in extending the tests, then I went for dinner,and enjoy the leisure time,
and then I found this mr is merged :sweat_smile: ,should I send another mr to update the code and extend the test?:)

Below is the draft left in the web page :sweat_smile: 

> So there are additional actions needed from the user to enable a specific translation?
 
Just as Milos said, users need to ask for  defining  the constraint and adding its support to tmt.hardware
This patch does not mean to overpass hardware.py .

> Sorry, no extra time as for now. But I'd say a very simple translation example and check with provision --dry would be 
> completely sufficient.

 yeah, I guess so, just asked to make sure.

**Comment by psss** [comment]
> Hi Petr, did you merge this by mistake? :)

I was just trying to push forward as much as possible before the release and thought that the missing stuff is not strictly blocking this. Also Miloš mentioned there are already some other pull requests with the planned documentation.
 
> I guess by "but I think it makes perfect sense to inspect the translations first, then fall back to hardcoded translations, then give up." Milos has the same opinion with you, ie, "let custom config surpass the default one in mrack.py" I already modify the code accordingly and halfway in extending the tests, then I went for dinner,and enjoy the leisure time, and then I found this mr is merged 😅

Oh, I see. Sorry for that, Lili! 🫣

> should I send another mr to update the code and extend the test?:)

Yes, please. If you have the tests extension prepared, let's definitely include it.

> I guess by "but I think it makes perfect sense to inspect the translations first, then fall back to hardcoded translations, then give up." Milos has the same opinion with you, ie, "let custom config surpass the default one in mrack.py"

I don't have a strong opinion on the exact order. But in general, yes it probably makes sense to give the user possibility to override hardcoded translations. So that approach sounds probably like that best option.

**Comment by skycastlelily** [comment]
> I was just trying to push forward as much as possible before the release
  
I guessed so:)

>Oh, I see. Sorry for that, Lili! 🫣

no problem,Petr:)

> Yes, please. If you have the tests extension prepared, let's definitely include it.

Here is the mr,including the tests and the refactor:)
https://github.com/teemtee/tmt/pull/3623
